src/hera/workflows/runner.py

```diff
@@ -64,30 +63,34 @@ def _parse(value, key, f):
         The parsed value.
 
     """
-    if _is_str_kwarg_of(key, f) or _is_artifact_loaded(key, f) or _is_output_kwarg(key, f):
+    if _is_str_kwarg_of(key, f):
+        return cast(type, _get_type(key, f))(value)
+    if _is_artifact_loaded(key, f) or _is_output_kwarg(key, f):
         return value
     try:
         return json.loads(value)
     except json.JSONDecodeError:
         return value
 
 
-def _is_str_kwarg_of(key: str, f: Callable):
-    """Check if param `key` of function `f` has a type annotation of a subclass of str."""
+def _get_type(key: str, f: Callable) -> Optional[type]:
     type_ = inspect.signature(f).parameters[key].annotation
-
     if type_ is inspect.Parameter.empty:
-        # Untyped args are interpreted according to json spec
-        # ie. we will try to load it via json.loads in _parse
-        return False
+        return None
     if get_origin(type_) is None:
-        return issubclass(type_, str)
-
+        return type_
     origin_type = cast(type, get_origin(type_))
     if origin_type is Annotated:
-        return issubclass(get_args(type_)[0], str)
+        return get_args(type_)[0]
+    return origin_type
+
 
-    return issubclass(origin_type, str)
+def _is_str_kwarg_of(key: str, f: Callable):
```

Comment:
Noticed we're a bit thin on function types in this file 😅 
```suggestion
def _is_str_kwarg_of(key: str, f: Callable) -> bool:
```